src/execution_plan/ops/op_merge.c

```diff
@@ -15,193 +14,289 @@
 static Record MergeConsume(OpBase *opBase);
 static void MergeFree(OpBase *opBase);
 
-static void _AddProperties(OpMerge *op, Record r, GraphEntity *ge, PropertyMap *props) {
-	for(int i = 0; i < props->property_count; i++) {
-		SIValue val = AR_EXP_Evaluate(props->values[i], r);
-		GraphEntity_AddProperty(ge, props->keys[i], val);
-		SIValue_Free(&val);
-	}
+//------------------------------------------------------------------------------
+// ON MATCH logic
+//------------------------------------------------------------------------------
+// Perform necessary index updates.
+static void _UpdateIndices(GraphContext *gc, Node *n) {
+	int label_id = Graph_GetNodeLabel(gc->g, n->entity->id);
+	if(label_id == GRAPH_NO_LABEL) return; // Unlabeled node, no need to update.
 
-	if(op->stats) op->stats->properties_set += props->property_count;
-}
+	Schema *s = GraphContext_GetSchemaByID(gc, label_id, SCHEMA_NODE);
+	if(!Schema_HasIndices(s)) return; // No indices, no need to update.
 
-/* Saves every entity within the query graph into the actual graph.
- * update statistics regarding the number of entities create and properties set. */
-static void _CommitNodes(OpMerge *op, Record r) {
-	int labelID;
-	Graph *g = op->gc->g;
+	Schema_AddNodeToIndices(s, n, true);
 
-	uint node_count = array_len(op->nodes_to_merge);
+}
 
-	// Start by creating nodes.
-	Graph_AllocateNodes(g, node_count);
+// Update the appropriate property on a graph entity.
+static void _UpdateProperty(Record r, GraphEntity *ge, EntityUpdateEvalCtx *update_ctx) {
+	SIValue new_value = AR_EXP_Evaluate(update_ctx->exp, r);
 
-	for(uint i = 0; i < node_count; i ++) {
-		NodeCreateCtx *node_ctx = &op->nodes_to_merge[i];
-		// Get blueprint of node to create
-		QGNode *node_blueprint = node_ctx->node;
+	// Try to get current property value.
+	SIValue *old_value = GraphEntity_GetProperty(ge, update_ctx->attribute_idx);
 
-		// Newly created node will be placed within given record.
-		Node *created_node = Record_GetNode(r, node_ctx->node_idx);
+	if(old_value == PROPERTY_NOTFOUND) {
+		// Add new property.
+		GraphEntity_AddProperty(ge, update_ctx->attribute_idx, new_value);
+	} else {
+		// Update property.
+		GraphEntity_SetProperty(ge, update_ctx->attribute_idx, new_value);
+	}
+}
 
-		Schema *schema = NULL;
+// Perform all ON MATCH updates for all matched records.
+static void _UpdateProperties(OpMerge *op, Record *records) {
```

Comment:
I originally wrote it that way, placing the Update op between Merge and the match stream ops, but this requires the match stream to buffer all records from the bound variable stream before executing, since Update is an eager operation.